### ISSUE TITLE
Kubelet does not handle valid DNS subdomain names/namespaces correctly

### ISSUE THREAD

**Issue by yujuhong** [GoogleCloudPlatform/kubernetes#4920]
This is forked from #3453.

Our documentation says that the name and the namespaces of a resource should be a valid DNS subdomain format: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/identifiers.md. 

Multiple places in Kubelet use `.` as a delimiter, even though dot could be used in a valid name or namespaces. For example,
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/config/config.go#l342
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/types.go#l79

A more concrete example would be that kubelet won't be able to run the two pods below at the same time as they all resort to the same name:
```
name: xxx, namespace: yyy.zzz  -> xxx.yyy.zzz
name: xxx.yyy, namespace: zzz  -> xxx.yyy.zzz
```

Note that the full pod name `<name>.<namespace>.<source>` is used to construct the container name. As @thockin suggested in #3453, kubelet should switch to using underscore `_` as the delimiter since it is not allowed in the DNS subdomain, but allowed in the docker container names. This change will not be backward compatible with existing container names.

Marking this P1 as this is a real bug.

**Comment by dchen1107** [comment]
We didn't fix this before is due to backward compatibility issue.

**Comment by yujuhong** [comment]
@dchen1107, it makes sense to me to fix this before v1.0 so that we could avoid hitting the compatibility issue further down the road.

**Comment by dchen1107** [comment]
Yes, we should fix it before 1.0. My point is that we should combine those destructive changes together and send one announcement to all users.

**Comment by dchen1107** [comment]
I am not sure if there are other destructive changes planned before 1.0, maybe some service related ones? cc/ @thockin

**Comment by thockin** [comment]
If this is just about bouncing containers I am not too concerned.

**Comment by yujuhong** [comment]
This is fixed by #5103
